File: 0101-0200/0191.Number_of_1_Bits.cpp
//0ms





class Solution {
public:
    int hammingWeight(int n)
    {
        int count = 0;
        while(n>0)
        {
            count += n&1;
            n >>= 1;
        }
        return count;
    }
};






class Solution {
public:
    int hammingWeight(int n) {
        int count = 0; // 1. Initialize a variable 'count' to 0. This counter will store the total number of '1' bits (set bits) found.

        while(n > 0) { // 2. Start a loop that continues as long as the integer 'n' is positive (i.e., there are still bits to examine).
            
            // 3. Check if the rightmost bit of 'n' is a 1 or a 0 using the bitwise AND operator (&).
            //    'n & 1' returns 1 if the last bit is 1, and 0 otherwise.
            //    This result (0 or 1) is immediately added to our running total 'count'.
            count += n & 1; 
            
            // 4. Shift all bits in 'n' one position to the right (right shift operator >>=).
            //    This effectively discards the last bit we just checked and moves the next bit into the rightmost position, preparing it for the next loop iteration.
            n >>= 1;
        }

        // 5. Once the loop terminates (when 'n' becomes 0), return the final accumulated 'count' of 1s.
        return count; 
    }
};






//0ms






class Solution {
public:
    int hammingWeight(int n) {
        return __builtin_popcount(n);
    }
};







class Solution {
public:
    int hammingWeight(int n) {
        // This is a direct, efficient solution leveraging a built-in compiler function.
        
        // 1. The `__builtin_popcount(n)` function is a GCC/Clang built-in intrinsic.
        //    It efficiently calculates the "population count" of the integer `n`. 
        //    "Population count" is a standard computing term for the number of '1' bits in the binary representation of an integer.
        
        // 2. The function delegates the entire task to this optimized intrinsic, 
        //    which is usually implemented directly by highly efficient hardware instructions (e.g., POPCNT on x86 processors).
        
        // 3. The function simply returns the value provided by the built-in call immediately.
        return __builtin_popcount(n);
    }
};

File: 0101-0200/0191.Number_of_1_Bits.py
#1ms



class Solution:
    def hammingWeight(self, n: int) -> int:
        res = 0

        for i in range(32):
            if (n >> i) & 1:
                res += 1

        return res









class Solution:
    def hammingWeight(self, n: int) -> int:
        # 1. Initialize 'res' to 0. This variable will count the total number of '1' bits found.
        res = 0

        # 2. Start a loop that iterates exactly 32 times (from i=0 to i=31).
        #    This covers all standard 32 bits of an integer in many common programming environments/constraints.
        for i in range(32):
            
            # 3. Check the status of the i-th bit from the right.
            #    (n >> i) shifts the number 'n' to the right by 'i' positions, moving the i-th bit to the rightmost position.
            #    The bitwise AND operation (& 1) then isolates that specific rightmost bit (result is 1 or 0).
            if (n >> i) & 1:
                
                # 4. If the result of the check is 1 (the bit is set), increment the result counter.
                res += 1

        # 5. After checking all 32 positions, return the final count of '1' bits.
        return res












#0ms





class Solution:
    def hammingWeight(self, n: int) -> int:
        binary=0
        result=0
        while n>0:
            result+=n&1
            n>>=1
        return result







class Solution:
    def hammingWeight(self, n: int) -> int:
        # Note: The 'binary = 0' variable is unused in this specific logic and can be safely ignored or removed.
        binary=0 

        # 1. Initialize 'result' to 0. This variable acts as our counter for the number of '1' bits found.
        result=0

        # 2. Start a loop that continues as long as the integer 'n' is greater than 0, ensuring all bits are checked.
        while n>0:
            # 3. Use the bitwise AND operator ('&') to check the rightmost (least significant) bit of 'n'.
            #    If the last bit is 1, 'n & 1' yields 1. If it's 0, it yields 0.
            #    This result is immediately added to our 'result' counter.
            result+=n&1
            
            # 4. Right shift 'n' by 1 position (n >>= 1). This operation discards the rightmost bit and moves 
            #    the next bit into the least significant position for the next iteration of the loop.
            n>>=1

        # 5. Once the loop finishes (when n reaches 0), return the final accumulated count of '1' bits.
        return result
